In [3]:
import pandas as pd
import numpy as np

In [4]:
data = pd.read_csv('adult.data.csv')
data.head()

age         workclass  fnlwgt  education  education-num  \
0   39         State-gov   77516  Bachelors             13   
1   50  Self-emp-not-inc   83311  Bachelors             13   
2   38           Private  215646    HS-grad              9   
3   53           Private  234721       11th              7   
4   28           Private  338409  Bachelors             13   

       marital-status         occupation   relationship   race     sex  \
0       Never-married       Adm-clerical  Not-in-family  White    Male   
1  Married-civ-spouse    Exec-managerial        Husband  White    Male   
2            Divorced  Handlers-cleaners  Not-in-family  White    Male   
3  Married-civ-spouse  Handlers-cleaners        Husband  Black    Male   
4  Married-civ-spouse     Prof-specialty           Wife  Black  Female   

   capital-gain  capital-loss  hours-per-week native-country salary  
0          2174             0              40  United-States  <=50K  
1             0             0              13  United-States  <=50K  
2             0             0              40  United-States  <=50K  
3             0             0              40  United-States  <=50K  
4             0             0              40           Cuba  <=50K

** 1. Сколько мужчин и женщин (признак sex) представлено в этом наборе данных? **

In [9]:
peopleCount = data['sex'].value_counts()
peopleCount.head()

Male      21790
Female    10771
Name: sex, dtype: int64

Ответ: Мужчин = 21790, Женщин = 10771

** 2. Каков средний возраст (признак age) женщин? **

In [40]:
female = data[data['sex'] == 'Female']
average = round(female['age'].mean(), 2) #округление до 2 знаков после точки
print(average)

36.86


Ответ: средний возраст женщин = 36.86

** 3. Какова доля граждан Германии (признак native-country)? **

In [55]:
obs =  data['native-country'].count() * 0.01
germanyPeople = data[data['native-country'] == 'Germany']['age'].count()
                   
print(round(germanyPeople / obs, 2))

0.42


Ответ: доля граждан Германии = 0.42%

** 4-5. Каковы средние значения и среднеквадратичные отклонения возраста тех, кто получает более 50K в год (признак salary) и тех, кто получает менее 50K в год? **

In [39]:
poor = data[(data['salary'] == '<=50K')]['age'].mean()
rich = data[(data['salary'] == '>50K')]['age'].mean()
print(round(poor, 2))
print(round(rich, 2))

poor = data[(data['salary'] == '<=50K')]['age'].std()
rich = data[(data['salary'] == '>50K')]['age'].std()
print(round(poor, 2))
print(round(rich, 2))

36.78
44.25
14.02
10.52


Ответ:
- те, кто получает более 50К в год: средние значения = 44.25, среднеквадратичные отклонения возраста = 10.52
- те, кто получает менее 50К в год: средние значения = 36.78, среднеквадратичные отклонения возраста = 14.02

**6. Правда ли, что люди, которые получают больше 50k, имеют как минимум высшее образование? (признак education – Bachelors, Prof-school, Assoc-acdm, Assoc-voc, Masters или Doctorate)**

In [50]:
pd.crosstab(data['salary'], data['education'])

education  10th  11th  12th  1st-4th  5th-6th  7th-8th  9th  Assoc-acdm  \
salary                                                                    
<=50K       871  1115   400      162      317      606  487         802   
>50K         62    60    33        6       16       40   27         265   

education  Assoc-voc  Bachelors  Doctorate  HS-grad  Masters  Preschool  \
salary                                                                    
<=50K           1021       3134        107     8826      764         51   
>50K             361       2221        306     1675      959          0   

education  Prof-school  Some-college  
salary                                
<=50K              153          5904  
>50K               423          1387

То же самое, но реализовано в виде списка:

In [108]:
list2 = ['Bachelors', 'Prof-school', 'Assoc-acdm', 'Assoc-voc', 'Masters или Doctorate']
data['edu'] = 0
data['edu'] = (data['education'].isin(list2)) * 1
pd.crosstab(data['salary'], data['education'])

education  10th  11th  12th  1st-4th  5th-6th  7th-8th  9th  Assoc-acdm  \
salary                                                                    
<=50K       871  1115   400      162      317      606  487         802   
>50K         62    60    33        6       16       40   27         265   

education  Assoc-voc  Bachelors  Doctorate  HS-grad  Masters  Preschool  \
salary                                                                    
<=50K           1021       3134        107     8826      764         51   
>50K             361       2221        306     1675      959          0   

education  Prof-school  Some-college  
salary                                
<=50K              153          5904  
>50K               423          1387

In [196]:
clever = data[(data['salary'] == '>50K')]['education'].value_counts()
print(clever)

Bachelors       2221
HS-grad         1675
Some-college    1387
Masters          959
Prof-school      423
Assoc-voc        361
Doctorate        306
Assoc-acdm       265
10th              62
11th              60
7th-8th           40
12th              33
9th               27
5th-6th           16
1st-4th            6
Name: education, dtype: int64


Ответ: судя по таблице, можно сделать вывод что нет, не правда. Существуют люди, которые получают >50K имея образование 10th, 11th, 12th и т.д.

** 7. Выведите статистику возраста для каждой расы (признак race) и каждого пола. Используйте groupby и describe. Найдите таким образом максимальный возраст мужчин расы Amer-Indian-Eskimo. **

In [66]:
ageStat = data.groupby('race')[['age', 'sex']].agg([np.min, np.max])
ageStat.head()

age          sex      
                   amin amax    amin  amax
race                                      
Amer-Indian-Eskimo   17   82  Female  Male
Asian-Pac-Islander   17   90  Female  Male
Black                17   90  Female  Male
Other                17   77  Female  Male
White                17   90  Female  Male

Ответ: максимальный возраст мужчин расы Amer-Indian-Eskimo = 82

** 8. Среди кого больше доля зарабатывающих много (>50K): среди женатых или холостых мужчин (признак marital-status)? Женатыми считаем тех, у кого marital-status начинается с Married (Married-civ-spouse, Married-spouse-absent или Married-AF-spouse), остальных считаем холостыми. **

In [109]:
obc = data[data['sex'] == 'Male']['sex'].count() * 0.01
married = data[(data['salary'] == '>50K') & (data['sex'] == 'Male') & (data['marital-status'].apply( \
                lambda x: 'married-' in x.lower()))]['sex'].count()
unmarried = data[(data['salary'] == '>50K') & (data['sex'] == 'Male') & (~data['marital-status'].apply( \
                lambda x: 'married-' in x.lower()))]['sex'].count()

print(round(married / obc, 2))
print(round(unmarried / obc, 2))

27.37
3.2


То же самое, но реализовано в виде списка:

In [139]:
list = ['Married-civ-spouse', 'Married-spouse-absent' , 'Married-AF-spouse']
data['is_married'] = 0
data['is_married'] = (data['marital-status'].isin(list)) * 1
round(pd.crosstab(data['salary'], data['is_married'], normalize = True) * 100, 2)

is_married      0      1
salary                  
<=50K       49.26  26.66
>50K         3.39  20.69

** 9. Какое максимальное число часов человек работает в неделю (признак hours-per-week)? Сколько людей работают такое количество часов и каков среди них процент зарабатывающих много? **

In [117]:
hoursCount = data['hours-per-week'].max() 
print(hoursCount)

count = data[data['hours-per-week'] == hoursCount].shape[0]
print(count)

richCount = data[data['hours-per-week'] == hoursCount]['salary'].value_counts(normalize='>50K')
print(richCount)

99
85
<=50K    0.705882
>50K     0.294118
Name: salary, dtype: float64


Ответ: максимальное число часов в неделю = 99, количество часов = 85, зарабатываюших много = 0.294118

** 10. Посчитайте среднее время работы (hours-per-week) зарабатывающих мало и много (salary) для каждой страны (native-country). **

In [140]:
round(pd.crosstab(data['native-country'], data['salary'], data['hours-per-week'], aggfunc='mean'), 2)

salary                      <=50K   >50K
native-country                          
?                           40.16  45.55
Cambodia                    41.42  40.00
Canada                      37.91  45.64
China                       37.38  38.90
Columbia                    38.68  50.00
Cuba                        37.99  42.44
Dominican-Republic          42.34  47.00
Ecuador                     38.04  48.75
El-Salvador                 36.03  45.00
England                     40.48  44.53
France                      41.06  50.75
Germany                     39.14  44.98
Greece                      41.81  50.62
Guatemala                   39.36  36.67
Haiti                       36.33  42.75
Holand-Netherlands          40.00    NaN
Honduras                    34.33  60.00
Hong                        39.14  45.00
Hungary                     31.30  50.00
India                       38.23  46.48
Iran                        41.44  47.50
Ireland                     40.95  48.00
Italy                       39.62  45.40
Jamaica                     38.24  41.10
Japan                       41.00  47.96
Laos                        40.38  40.00
Mexico                      40.00  46.58
Nicaragua                   36.09  37.50
Outlying-US(Guam-USVI-etc)  41.86    NaN
Peru                        35.07  40.00
Philippines                 38.07  43.03
Poland                      38.17  39.00
Portugal                    41.94  41.50
Puerto-Rico                 38.47  39.42
Scotland                    39.44  46.67
South                       40.16  51.44
Taiwan                      33.77  46.80
Thailand                    42.87  58.33
Trinadad&Tobago             37.06  40.00
United-States               38.80  45.51
Vietnam                     37.19  39.20
Yugoslavia                  41.60  49.50